src/ui/suspense/frameCache.ts

```diff
@@ -7,12 +7,16 @@ export const {
   getValueSuspense: getFramesSuspense,
   getValueAsync: getFramesAsync,
   getValueIfCached: getFramesIfCached,
-} = createGenericCache<[pauseId: PauseId], Frame[] | undefined>(
+  useGetValue: useGetFrames,
+} = createGenericCache<[pauseId: PauseId | undefined], Frame[] | undefined>(
   async pauseId => {
+    if (!pauseId) {
+      return undefined;
+    }
```

Comment:
Under what condition would this method be called without a pause id? That feels suspicious to me.

We should just not call this method if there's no pause id.

Reply:
I've added `useGetValue(...)` to the generic cache (which is used as `useGetFrames(...)` in `CommandBar.tsx`) - since that is a hook, we can't bail before calling it, even if calling it makes little sense in that case.

Reply:
Why do we need `useGetValue` ?

Reply:
We don't strictly need it but it's convenient in `CommandBar.tsx` and `Toolbar.tsx`: in both cases we want to enable some buttons or add a badge to one of them if there are frames for the current point. Doing this with a suspense mechanism would be less straightforward.

Reply:
> Doing this with a suspense mechanism would be less straightforward.

Why?

I feel uneasy with us reinventing an alternative Suspense model here. Feels like we're complicating things unnecessary, and it introduces extra constraints like the "can't be called conditionally" thing you. mentioned above.

Reply:
> Doing this with a suspense mechanism would be less straightforward.

I think maybe there's something here that's being misunderstood. Unless I'm missing some important context, I think Suspense should be a good fit for the command bar and toolbar components too. Maybe you can elaborate on what the complicated part is?

Reply:
Maybe I'm just blind. Imagine you're doing something like this:
```
function CommandBar({ pauseId }) {
  const { frames } = useGetFrames(pauseId);
  const disabled = !frames?.length;
  return (
    <Button stuff={...} disabled={disabled} />
    <Button stuff={...} />
    <Button stuff={...} disabled={disabled} />
  );
```
How would you do that with `getFramesSuspense()`? Maybe something like this:
```
function CommandBar({ pauseId }) {
  return (
    <Suspense fallback={<Button stuff={...} disabled={true} />}>
      <SuspendingButton pauseId={pauseId} stuff={...} />
    </Suspense>
    <Button stuff={...} />
    <Suspense fallback={<Button stuff={...} disabled={true} />}>
      <SuspendingButton pauseId={pauseId} stuff={...} />
    </Suspense>
}

function SuspendingButton({ pauseId, ...stuff }) {
  const frames = getFramesSuspense(pauseId);
  const disabled = !frames?.length;
  return <Button stuff={...stuff} disabled={disabled} />;
}
```
And that looks a lot less straightforward to me. But maybe you know a cleaner way to write this with suspense.

Reply:
Ah, thanks for elaborating. The transition and deferred APIs can be used for this sort of thing.

```js
function CommandBar({ pauseId }) {
  const deferredPauseId = useDeferredValue(pauseId);
  const isPending = deferredPauseId !== pauseId;

  const { frames } = deferredPauseId != null ? getFramesSuspense(pauseId) : null;
  const hasFrames = !!frames?.length;

  const disabled = isPending || !hasFrames;

  return (
    <Button stuff={...} disabled={disabled} />
    <Button stuff={...} />
    <Button stuff={...} disabled={disabled} />
  );
```

Reply:
Interesting, thanks. I'll have to read up on `useDeferredValue()` later to understand this.

Reply:
Would you be interested in pairing on that change? so we can avoid adding a Suspense-alternative like `useGetValue`?

Reply:
I have just pushed an improved hook that doesn't violate the rules and is better at avoiding unnecessary rerenders.
I've also experimented with your solution above but `useDeferredValue()` didn't work as I expected, in particular on the very first render `useDeferredValue(pauseId)` will return `pauseId`, so the component will suspend (instead of rendering disabled buttons).

Reply:
> on the very first render `useDeferredValue(pauseId)` will return pauseId, so the component will suspend (instead of rendering disabled buttons).

Isn't that expected? If there's no data to show (even in an "outdated but this is updating" way) then it should suspend, no?

Reply:
> Isn't that expected?

It's not what *I* expected. The original implementation would always render disabled buttons while frames are being fetched and I wanted to keep that behavior. If we want to change it so that the buttons keep their old disabled state until the frames are fetched, transitions are obviously what we need. But I wanted to discuss how to implement the old behavior with the suspense cache.

Reply:
Okay. Fair enough.

What I was trying to say is just that...that's how Suspense works. When content is loading the first time, React shows the fallback UI (which is generally some kind of spinner or "_loading_" text). Knowing this, you can design the UI to "fallback" however you'd like.

My point is that this isn't a bug or a problem. It's just how Suspense is meant to work.

Reply:
So would you agree that the `useGetValue()` hook is useful if we want to keep the old behavior?

Reply:
I still think the `useGetValue` abstraction is unnecessary and that we should use Suspense as it works out of the box rather than work around it by adding extra state and effects, but I understand that we don't share that opinion.

I wrote the code snippet above under the assumption that we were talking about _updates_ (because those are a little more complicated until you understand how transitions work).

I'm probably not going to think about this PR too much until the points stuff is done and merged because it's kind of sucking up all of my mental energy at the moment. Maybe we can chat about this again on Monday?